api/alertmanager/v1/alertmanager/api.go

```diff
@@ -0,0 +1,280 @@
+	if err != nil {
+		return nil, err
+	}
+
+	var ss *ServerStatus
+	err = json.Unmarshal(body, &ss)
+
+	return ss, err
+}
+
+type AlertAPI interface {
+	// List all the active alerts.
+	List(ctx context.Context) ([]*model.Alert, error)
+	// Push a list of alerts into the Alertmanager.
+	Push(ctx context.Context, alerts ...*model.Alert) error
+}
+
+// NewAlertAPI returns a new AlertAPI for the client.
+func NewAlertAPI(c api.Client) AlertAPI {
+	return &httpAlertAPI{client: apiClient{c}}
+}
+
+type httpAlertAPI struct {
+	client api.Client
+}
+
+func (h *httpAlertAPI) List(ctx context.Context) ([]*model.Alert, error) {
+	u := h.client.URL(epAlerts, nil)
+
+	req, _ := http.NewRequest("GET", u.String(), nil)
+
+	_, body, err := h.client.Do(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	var alts []*model.Alert
+	err = json.Unmarshal(body, &alts)
+
+	return alts, err
+}
+
+func (h *httpAlertAPI) Push(ctx context.Context, alerts ...*model.Alert) error {
+	u := h.client.URL(epAlerts, nil)
+
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(alerts); err != nil {
+		return err
+	}
+
+	req, _ := http.NewRequest("POST", u.String(), &buf)
+
+	_, _, err := h.client.Do(ctx, req)
+	return err
+}
+
+// SilenceAPI provides bindings the Alertmanager's silence API.
+type SilenceAPI interface {
+	// Get returns the silence associated with the given ID.
+	Get(ctx context.Context, id uint64) (*model.Silence, error)
+	// Set updates or creates the given silence and returns its ID.
+	Set(ctx context.Context, sil *model.Silence) (uint64, error)
+	// Del deletes the silence with the given ID.
+	Del(ctx context.Context, id uint64) error
+	// List all silences of the server.
+	List(ctx context.Context) ([]*model.Silence, error)
+}
+
+// NewSilenceAPI returns a new SilenceAPI for the client.
+func NewSilenceAPI(c api.Client) SilenceAPI {
+	return &httpSilenceAPI{client: apiClient{c}}
+}
+
+type httpSilenceAPI struct {
+	client api.Client
+}
+
+func (h *httpSilenceAPI) Get(ctx context.Context, id uint64) (*model.Silence, error) {
+	u := h.client.URL(epSilence, map[string]string{
+		"id": strconv.FormatUint(id, 10),
+	})
+
+	req, _ := http.NewRequest("GET", u.String(), nil)
+
+	_, body, err := h.client.Do(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	var sil model.Silence
+	err = json.Unmarshal(body, &sil)
+
+	return &sil, err
+}
+
+func (h *httpSilenceAPI) Del(ctx context.Context, id uint64) error {
+	u := h.client.URL(epSilence, map[string]string{
+		"id": strconv.FormatUint(id, 10),
+	})
+
+	req, _ := http.NewRequest("DELETE", u.String(), nil)
+
+	_, _, err := h.client.Do(ctx, req)
+	return err
+}
+
+func (h *httpSilenceAPI) Set(ctx context.Context, sil *model.Silence) (uint64, error) {
+	var (
+		u      *url.URL
+		method string
+	)
+
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(sil); err != nil {
+		return 0, err
+	}
+
+	// Talk to different endpoints depending on whether its a new silence or not.
+	if sil.ID != 0 {
```

Comment:
I'm of the opinion that `sil` ought to be a value (not a pointer). Imagine if some goroutine also had the `sil` address and was messing with its properties during the `Encode` above; it would leave the request in an inconsistent state (i.e. not in the user's intention at the point that `Set` was called). Agreed?